kolibri/plugins/coach/assets/src/composables/useQuizResources.js

```diff
@@ -0,0 +1,138 @@
+import { validateObject } from 'kolibri.utils.objectSpecs';
+import { ContentNodeResource } from 'kolibri.resources';
+import { ContentNodeKinds } from 'kolibri.coreVue.vuex.constants';
+import useFetchTree from './useFetchTree';
+import { QuizExercise } from './quizCreationSpecs.js';
+
+/**
+ * @typedef {Object} QuizResourcesConfig
+ * @property { computed <string|null|undefined> } topicId - The id of the root node to fetch the
+ * children for
+ */
+
+/**
+ * @module useQuizResources
+ * @param {QuizResourcesConfig} config
+ */
+export default function useQuizResources({ topicId } = {}) {
+  const params = {
+    kind_in: [ContentNodeKinds.EXERCISE, ContentNodeKinds.TOPIC],
+    include_coach_content: true,
+  };
+
+  // Initialize useFetchTree methods with the given topicId computed property and params
+  const { topic, fetchTree, fetchMore, hasMore, loading: treeLoading } = useFetchTree({
+    topicId,
+    params,
+  });
+
+  /** @type {ref<ExerciseResource[]>} All resources which have been fetched that are the children of
+   * the given topicId annotated with assessment metadata */
+  const _resources = ref([]);
+
+  /** @type {ref<Boolean>} Whether we are currently fetching/processing the child nodes */
+  const _loading = ref(false);
+
+  /**
+   * Annotates the child TOPIC nodes with the number of assessments that are contained within them
+   * @param {string[]} topicsIds - The list of topics IDs to fetch descendant counts for
+   * @affects _resources - The topicIds passed here will have their `num_assessments` property
+   *    added to the corresponding Topic within _resources
+   * @returns {Promise<null>} - A promise that resolves when the annotations have been made and
+   *   the_resources have been updated, filtering out all topics which do not have assessments
+   */
+  async function annotateTopicsWithDescendantCounts(topicIds = []) {
+    return ContentNodeResource.fetchDescendantsAssessments(topicIds)
+      .then(({ data: topicsWithDescendantCounts }) => {
+        //console.log(topicsWithDescendantCounts);
+        const childrenWithAnnotatedTopics = get(_resources)
+          .map(node => {
+            // We'll map so that the topics are updated in place with the num_assessments, others
+            // are left as-is
+            if (node.kind === ContentNodeKinds.TOPIC) {
+              const topic = topicsWithDescendantCounts.find(t => t.id === node.id);
+              if (topic) {
+                node.num_assessments = topic.num_assessments;
+              }
+              if (!validateObject(node, QuizExercise)) {
+                console.warn('Topic node was not a valid QuizExercise after annotation:', node);
+              }
+            }
+            return node;
+          })
+          .filter(node => {
+            // Only keep topics which have assessments in them to begin with
+            if (node.kind === ContentNodeKinds.TOPIC) {
+              if (node.num_assessments > 0) {
```

Comment:
Could reduce nesting slightly here with an early return strategy, and by just returning the value of this comparison:
```
if (node.kind !== ContentNodeKinds.TOPIC) {
  return true;
}
return node.num_assessments > 0;
```